alert/log/search.go

```diff
@@ -0,0 +1,146 @@
+package alertlog
+
+import (
+	"context"
+	"database/sql"
+	"github.com/target/goalert/permission"
+	"github.com/target/goalert/search"
+	"github.com/target/goalert/validation"
+	"github.com/target/goalert/validation/validate"
+	"text/template"
+
+	"github.com/lib/pq"
+	"github.com/pkg/errors"
+)
+
+// SearchOptions contains criteria for filtering alert logs.
+type SearchOptions struct {
+	// FilterAlertIDs restricts the log entries belonging to specific alertIDs only.
+	FilterAlertIDs []int
+
+	// Limit restricts the maximum number of rows returned. Default is 15.
+	Limit int
+
+	After SearchCursor `json:"a,omitempty"`
+}
+
+type SearchCursor struct {
+	ID int `json:"i,omitempty"`
+}
+
+var searchTemplate = template.Must(template.New("search").Parse(`
+	SELECT
+		a.id, 
+		a.alert_id,
+		a.timestamp,
+		a.event,
+		a.message,
+		a.sub_type,
+		a.sub_user_id,
+		u.name,
+		a.sub_integration_key_id,
+		i.name,
+		a.sub_hb_monitor_id,
+		hb.name,
+		a.sub_channel_id,
+		nc.name,
+		a.sub_classifier,
+		a.meta
+	FROM alert_logs a
+	{{if .FilterAlertIDs }}
+		LEFT JOIN alerts ON alerts.id = a.alert_id
+	{{end}}
+	LEFT JOIN users u ON u.id = a.sub_user_id
+	LEFT JOIN integration_keys i ON i.id = a.sub_integration_key_id
+	LEFT JOIN heartbeat_monitors hb ON hb.id = a.sub_hb_monitor_id 
+	LEFT JOIN notification_channels nc ON nc.id = a.sub_channel_id
+	WHERE TRUE
+	{{if .FilterAlertIDs}}
+		AND a.alert_id = ANY(:alertIDs)
+	{{end}}
+	{{- if .After.ID}}
+		AND (a.id < :afterID)
+	{{- end}}
+	ORDER BY a.id DESC
+	LIMIT {{.Limit}}
+`))
+
+type renderData SearchOptions
+
+func (opts renderData) Normalize() (*renderData, error) {
+	if opts.Limit == 0 {
+		opts.Limit = search.DefaultMaxResults
+	}
+
+	var err error
+	if len(opts.FilterAlertIDs) <= 0 {
```

Comment:
These don't need to be required, the query will work fine without and filter IDs. We impose a max limit so the cost is already bounded.

The only thing we should validate is that the length of `FilterAlertIDs` isn't astronomical (since we send the list to the DB). Maybe limit it to 50 or something beyond what is realistic, but well below an amount that could cause problems.